src/main/java/io/cdap/plugin/google/sheets/sink/GoogleSheetsRecordWriter.java

```diff
@@ -0,0 +1,373 @@
+      TimeUnit.SECONDS);
+
+    // we should guarantee that at least one task former was called finally
+    try {
+      formerScheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+      formerScheduledExecutorService.submit(new TasksFormer(false)).get();
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e.getMessage());
+    }
+
+    // wait for worker threads completion
+    writeService.shutdown();
+    writeService.awaitTermination(flushTimeout, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Task that forms batches of requests and schedules tasks for writing the batches.
+   */
+  public class TasksFormer implements Callable {
+    private final boolean rerun;
+
+    public TasksFormer(boolean rerun) {
+      this.rerun = rerun;
+    }
+
+    @Override
+    public Object call() throws Exception {
+      queueSemaphore.acquire();
+      try {
+        Map<String, Queue<FlatternedRowsRequest>> groupedRequests = new HashMap<>();
+
+        // pick up groups of elements
+        FlatternedRowsRequest element;
+        while ((element = recordsQueue.poll()) != null) {
+          String spreadSheetName = element.getSpreadSheetName();
+
+          if (!groupedRequests.containsKey(spreadSheetName)) {
+            groupedRequests.put(spreadSheetName, new ConcurrentLinkedQueue<>());
+          }
+          groupedRequests.get(spreadSheetName).add(element);
+        }
+
+        // form batches of requests
+        List<RecordsBatch> recordsToSort = new ArrayList<>();
+        for (Map.Entry<String, Queue<FlatternedRowsRequest>> spreadSheetEntry : groupedRequests.entrySet()) {
+          String spreadSheetName = spreadSheetEntry.getKey();
+          Queue<FlatternedRowsRequest> queue = spreadSheetEntry.getValue();
+          List<FlatternedRowsRequest> buffer = new ArrayList<>();
+
+          while (!queue.isEmpty()) {
+            buffer.add(queue.poll());
+            if (buffer.size() == googleSheetsSinkConfig.getMaxBufferSize()) {
+              recordsToSort.add(new RecordsBatch(buffer, spreadSheetName));
+              buffer = new ArrayList<>();
+            }
+          }
+          if (!buffer.isEmpty()) {
+            recordsToSort.add(new RecordsBatch(buffer, spreadSheetName));
+          }
+        }
+
+        // sort groups by size of records
+        Collections.sort(recordsToSort);
+
+        // send biggest groups to threads, else back to queue
+        for (RecordsBatch recordsBatch : recordsToSort) {
+          if (stopSignal) {
+            if (threadsSemaphore.tryAcquire(flushTimeout, TimeUnit.SECONDS)) {
+              // create new thread
+              writeService.submit(
+                new MessagesProcessor(recordsBatch));
+            } else {
+              throw new RuntimeException(
+                String.format("It is not possible schedule of records batch during '%d' seconds.", flushTimeout));
```

Comment:
Timeout '%d' exceeded when trying to schedule batch records for execution.

Reply:
done